src/clojure/nrepl/middleware/session.clj

```diff
@@ -216,60 +292,80 @@
      (try-stop-thread t))))
 
 (defn session-exec
-  "Takes a session id and returns a maps of three functions meant for interruptible-eval:
-   * :exec, takes an id (typically a msg-id), a thunk and an ack runnables (see #'default-exec for ampler
-     context). Executions are serialized and occurs on a single thread.
-   * :interrupt, takes an id and tries to interrupt the matching execution (submitted with :exec above).
-     A nil id is meant to match the currently running execution. The return value can be either:
-     :idle (no running execution), the interrupted id, or nil when the running id doesn't match the id argument.
-     Upon successful interruption the backing thread is replaced.
-   * :close, terminates the backing thread."
-  [id]
-  (let [queue (LinkedBlockingQueue.)
-        running (atom nil)
-        thread (atom nil)
-        main-loop #(try
-                     (loop []
-                       (let [[exec-id ^Runnable r ^Runnable ack] (.take queue)]
-                         (reset! running exec-id)
-                         (when (try
-                                 (.run r)
-                                 (compare-and-set! running exec-id nil)
-                                 (finally
-                                   (compare-and-set! running exec-id nil)))
-                           (some-> ack .run)
-                           (recur))))
-                     (catch InterruptedException _e))
-        spawn-thread #(doto (Thread. main-loop (str "nREPL-session-" id))
+  "Takes a session and returns a maps of three functions meant for
+  interruptible-eval:
+  - `:exec` - takes an id (typically a msg-id), a thunk and an ack runnables,
+    and an optional messsage (see `make-ephemeral-exec-fn` for ampler context).
+    Executions are serialized and occurs on a single thread.
+  - `:interrupt` - takes an id and tries to interrupt the matching execution
+    (submitted with :exec above). A nil id is meant to match the currently
+    running execution. The return value can be either: `:idle` (no running
+    execution), the interrupted id, or nil when the running id doesn't match the
+    id argument. Upon successful interruption the backing thread is replaced.
+  - `:close` - terminates the backing thread."
+  [session]
```

Comment:
Why did you decide to switch from session id to session here?

Reply:
Because session is what contains all the saved dynamic bindings, and this function establishes those bindings for the code being run.

Reply:
Yeah, I get this - I was mostly wondering if it might break some client calling `session-exec` directly for some reason, as it's public API. However, this seems fairly unlikely to me.

Reply:
Github search shows nothing. And public functions in the namespace that defines a middleware are not really part of the API, only the middleware itself is.